### PR TITLE
[FIX] mail: mass mailing to same mail adress

### PR DESCRIPTION
**Pull request by yosa-odoo**
Steps to repoduce:
- Accounting > Customers > Invoices:
	 select several invoices to send
- Action > Send & Print > (deselect Print) > Send & Print

Issue:
- It sends only one invoice per company

Cause:
- the mail_compose_message sets the status of an email as `cancel` when a mail has already been sent to a specific adress mail in the batch

Solution:
- If the use of mass mailing is document-based (e.g.: sending multiple invoices) it will allow to send multiple emails to the same adress

opw-2775121

### CODE REVIEW

**Comment by yosa-odoo** [comment]
@fw-bot ignore

**Comment by yosa-odoo** [comment]
The use of context key will be used for stable. For Master, a field will be created for `mail_compose_message`.

**Comment by wamdam** [comment]
Doing this will hit you as soon as one customer sets a static subject in the mail template.

File: addons/mail/wizard/mail_compose_message.py [hunk 1]
```diff
@@ -472,6 +472,8 @@ def _process_state(self, mail_values_dict):
         blacklist_ids = self._get_blacklist_record_ids(mail_values_dict)
         optout_emails = self._get_optout_emails(mail_values_dict)
         done_emails = self._get_done_emails(mail_values_dict)
+        # in case of an invoice e.g.
+        document_based = self.env.context.get('document_based')
```
Comment (tde-banana-odoo) [inline addons/mail/wizard/mail_compose_message.py:476]: Would make a prefixed context key, like ``mailing_document_based``

File: addons/account/wizard/account_invoice_send.py [hunk 2]
```diff
@@ -116,7 +116,7 @@ def _compute_invoice_without_email(self):
     def _send_email(self):
         if self.is_email:
             # with_context : we don't want to reimport the file we just exported.
-            self.composer_id.with_context(no_new_invoice=True, mail_notify_author=self.env.user.partner_id in self.composer_id.partner_ids)._action_send_mail()
+            self.composer_id.with_context(no_new_invoice=True, mail_notify_author=self.env.user.partner_id in self.composer_id.partner_ids, document_based=True)._action_send_mail()
```
Comment (tde-banana-odoo) [inline addons/account/wizard/account_invoice_send.py:119]: Linting: could be multi-lined to ease reading (beginning to explode outside of my screen ^^)

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for the PR ! Would love to have an unit test for that ... could be done in test_mail, we already have some tests for the composer. Just try a mass composer sending with two documents having the same email, check that either we have 1 outgoing email, either 2, depending on the key :) .

cc @qdp-odoo 

Cheers !

**Comment by yosa-odoo** [comment]
@tde-banana-odoo coucou
I created a unit test for that matter. I'm just checking that we have two outgoing mails with the keys since the duplicate check is verified juste before https://github.com/odoo/odoo/blob/8a339bc33b6061d3f2461ed79d5e7a92a0268370/addons/test_mail/tests/test_mail_composer.py#L615

File: addons/mail/wizard/mail_compose_message.py [hunk 3]
```diff
@@ -494,7 +496,7 @@ def _process_state(self, mail_values_dict):
             elif optout_emails and mail_to in optout_emails:
                 mail_values['state'] = 'cancel'
                 mail_values['failure_type'] = 'mail_optout'
-            elif done_emails and mail_to in done_emails:
+            elif done_emails and mail_to in done_emails and not mailing_document_based:
```
Comment (william-andre) [inline addons/mail/wizard/mail_compose_message.py:499]: Shouldn't it depend on `record_id` when it is `mailing_document_based`?
Actually I don't understand why we need the context key if the record is different? (maybe to ensure that we are not breaking anything in stable?)

Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py:499]: Note that we already loop on res_ids, meaning normally you will never have twice the same document in the loop.

Concerning the context key: because in mass mailing application, you don't want to mail a given email address several times (because mailing lists could contain thousands of emails with duplicated). But when sending invoices in batch, you want to email the same address several times if it has several invoices.

Basically we have two behavior: marketing (don't spam) and business (allow spam).

**Comment by tde-banana-odoo** [comment]
Hello @yosa-odoo ,

Looking at it again, thanks for update ! Don't you think we should do this move for all models, or at least for Invoices / SO / PO ? If we consider it as a bug for invoices, this is probably the same for those models. Not sure for other models, this is a functional change so always difficult to balance.

In master I think we can consider that "only having mailing_id set on the composer triggers the marketing behavior by default". Maybe to confirm with Lna on the task.

cc @william-andre in case you have some thoughts here, as we already had to discuss whether something is a bugfix or a feature, and as it impacts invoice model ^^ .

**Comment by yosa-odoo** [comment]
>Don't you think we should do this move for all models, or at least for Invoices / SO / PO ?
I'm gonna check it out.

>In master I think we can consider that "only having mailing_id set on the composer triggers the marketing behavior by default".
I'll do the check right after. I was using a computed field which would check the model if `mailing.contact` -> marketing. But maybe there other cases/models I'm not thinking of.

**Comment by yosa-odoo** [comment]
It does not seem that for SO and PO, there is the possibility to select multiple elements and send them as email (as it is for invoices).

If I'm wrong, the change of adding the context key would have to be made here

For Sales:
https://github.com/odoo/odoo/blob/e9af6096167392c2d3605389367eba9df1b68c50/addons/sale/wizard/mail_compose_message.py#L10-L13

For the Purchase:
https://github.com/odoo/odoo/blob/e9af6096167392c2d3605389367eba9df1b68c50/addons/purchase/models/mail_compose_message.py#L10-L13

**Comment by wamdam** [comment]
@yosa-odoo There is. Just create a mail template, click on "Create context action" and use this button.
The _action_send_mail method should work as expected on any model with very small exceptions (like mailing lists, contacts).

**Comment by yosa-odoo** [comment]
@tde-banana-odoo Do you need other tests for `purchase` and `sales` or the general test using the context key is enough?

**Comment by yosa-odoo** [comment]
d

File: addons/test_mail/tests/test_mail_composer.py [hunk 4]
```diff
@@ -659,6 +659,29 @@ def test_mail_composer_wtpl_complete(self):
         self.assertEqual(set(message.attachment_ids.mapped('res_id')), set(self.test_record.ids))
         self.assertTrue(all(attach not in message.attachment_ids for attach in attachs), 'Should have copied attachments')
 
+    def test_mail_composer_document_based(self):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:662]: Could you move test definition up ? As tests are executed alphabetically in a given class, better order them alphabetically so that test execution order matches definition order. That way when you have several issues, you can follow console errors in the file ^^ .

File: addons/test_mail/tests/test_mail_composer.py [hunk 5]
```diff
@@ -659,6 +659,29 @@ def test_mail_composer_wtpl_complete(self):
...
+        """ Tests a document-based mass mailing with the same address mails
+        This should be allowed and not considered as duplicate in this context
+        """
+        attachment_data = self._generate_attachments_data(2)
+        email_to_1 = 'test.to.1@test.example.com'
+        self.template.write({
+            'auto_delete': False,  # keep sent emails to check content
+            'attachment_ids': [(0, 0, a) for a in attachment_data],
+            'email_to': '%s, %s' % (email_to_1, email_to_1),
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_composer.py:671]: Shouldn't we set the same partner / email on all test_records to clearly mimic the "invoice with same recipient" use case ? Not sure we are not mixing several tests here.

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for followup ! I don't think we have to add tests in each module (testing phase + unit test should be sufficient). I have some remarks on tests while passing by :) .

Cheers !

**Comment by yosa-odoo** [comment]
@tde-banana-odoo I changed the test, so only two mails (and not four) are sent.

**Comment by yosa-odoo** [comment]
@fw-bot  up to saas-15.2

**Comment by tde-banana-odoo** [comment]
@yosa-odoo Better let the PR go until master, then change the implementation in master. Anyway as we have other stuff planned in composer in master, I think we will retake the master branch in a more global change :) .

**Comment by yosa-odoo** [comment]
@tde-banana-odoo I just created a branch and cherry-pick this commit and the forward-port up to saas-15.2.
https://github.com/odoo/odoo/pull/88125

Should I close it then?

**Comment by tde-banana-odoo** [comment]
Forward port are disabled due to the freeze. It will resume after next Monday.

**Comment by william-andre** [comment]
> Forward port are disabled due to the freeze. It will resume after next Monday.

Nothing to do with the freeze, the PR is merged and the setting cannot be changed after that.
https://github.com/odoo/odoo/pull/85918#issuecomment-1063857180

**Comment by tde-banana-odoo** [comment]
@william-andre I was speaking about the forward branches not being created currently.
